numpy/lib/function_base.py

```diff
@@ -4014,122 +4135,234 @@ def _quantile_is_valid(q):
-    # round fractional indices according to interpolation method
-    if interpolation == 'lower':
-        indices = floor(indices).astype(intp)
-    elif interpolation == 'higher':
-        indices = ceil(indices).astype(intp)
-    elif interpolation == 'midpoint':
-        indices = 0.5 * (floor(indices) + ceil(indices))
-    elif interpolation == 'nearest':
-        indices = around(indices).astype(intp)
-    elif interpolation == 'linear':
-        pass  # keep index as fraction and interpolate
-    else:
-        raise ValueError(
-            "interpolation can only be 'linear', 'lower' 'higher', "
-            "'midpoint', or 'nearest'")
 
+def _quantile(
+        arr: np.array,
+        quantiles: np.array,
+        axis: int = -1,
+        interpolation=QuantileInterpolation.INCLUSIVE,
+        out=None,
+):
+    """
+    Private function that doesn't support extended axis or keepdims.
+    These methods are extended to this function using _ureduce
+    See nanpercentile for parameter usage
+    It computes the quantiles of the array for the given axis.
+    A linear interpolation is performed based on the interpolation_type.
+
+    By default, the interpolation is QuantileInterpolation.INCLUSIVE where
+    alpha == beta == 1 which performs the 7th method of Hyndman&Fan.
+    With QuantileInterpolation.MEDIAN_UNBIASED we get  alpha == beta == 1/3
+    thus the 8th method of Hyndman&Fan.
+    """
+    # --- Setup
+    # ufuncs cause 0d array results to decay to scalars (see gh-13105), which
+    # makes them problematic for __setitem__ and attribute access. As a
+    # workaround, we call this on the result of every ufunc on a possibly-0d
+    # array.
+    not_scalar = np.asanyarray
+    values_count = arr.shape[axis]
+    if values_count == 0:
+        return np.NAN
+    if values_count == 1:
+        result = np.take(arr, indices=0, axis=axis)
+        if np.issubdtype(arr.dtype, np.inexact) and np.isnan(result):
+            warnings.warn("All-NaN slice encountered", RuntimeWarning,
+                          stacklevel=4)
+        return np.broadcast_to(result, (quantiles.size,) + result.shape)
     # The dimensions of `q` are prepended to the output shape, so we need the
     # axis being sampled from `ap` to be first.
-    ap = np.moveaxis(ap, axis, 0)
-    del axis
-
-    if np.issubdtype(indices.dtype, np.integer):
-        # take the points along axis
-
-        if np.issubdtype(a.dtype, np.inexact):
-            # may contain nan, which would sort to the end
-            ap.partition(concatenate((indices.ravel(), [-1])), axis=0)
-            n = np.isnan(ap[-1])
-        else:
-            # cannot contain nan
-            ap.partition(indices.ravel(), axis=0)
-            n = np.array(False, dtype=bool)
-
-        r = take(ap, indices, axis=0, out=out)
-
+    arr = np.moveaxis(arr, axis, destination=0)
+    valid_values_count = np.broadcast_to(not_scalar(values_count),
+                                         arr.shape[1:])
```

Comment:
```suggestion
    valid_values_count = np.broadcast_to(values_count, arr.shape[1:], 
                                         subok=True)
```
No need to explicitly call `asanyarray` here, just let `broadcast_to` do it under the hood.